sklearn/pipeline.py

```diff
@@ -135,40 +218,59 @@ def _final_estimator(self):
 
     # Estimator interface
 
-    def _pre_transform(self, X, y=None, **fit_params):
-        fit_params_steps = dict((step, {}) for step, _ in self.steps)
+    def _fit(self, X, y=None, **fit_params):
+        fit_params_steps = dict((name, {}) for name, step in self.steps
+                                if step is not None)
         for pname, pval in six.iteritems(fit_params):
             step, param = pname.split('__', 1)
             fit_params_steps[step][param] = pval
         Xt = X
         for name, transform in self.steps[:-1]:
-            if hasattr(transform, "fit_transform"):
+            if transform is None:
+                pass
+            elif hasattr(transform, "fit_transform"):
                 Xt = transform.fit_transform(Xt, y, **fit_params_steps[name])
             else:
                 Xt = transform.fit(Xt, y, **fit_params_steps[name]) \
                               .transform(Xt)
         return Xt, fit_params_steps[self.steps[-1][0]]
 
     def fit(self, X, y=None, **fit_params):
-        """Fit all the transforms one after the other and transform the
+        """Fit the model
+
+        Fit all the transforms one after the other and transform the
         data, then fit the transformed data using the final estimator.
 
         Parameters
         ----------
         X : iterable
             Training data. Must fulfill input requirements of first step of the
             pipeline.
+
         y : iterable, default=None
             Training targets. Must fulfill label requirements for all steps of
             the pipeline.
+
+        **fit_params : dict of string -> object
+            Parameters passed to the ``fit`` method of each step, where
+            each parameter name is prefixed such that parameter ``p`` for step
+            ``s`` has key ``s__p``.
+
+        Returns
+        -------
+        self : Pipeline
+            This estimator
         """
-        Xt, fit_params = self._pre_transform(X, y, **fit_params)
+        self._validate_steps()
+        Xt, fit_params = self._fit(X, y, **fit_params)
         self.steps[-1][-1].fit(Xt, y, **fit_params)
         return self
 
     def fit_transform(self, X, y=None, **fit_params):
-        """Fit all the transforms one after the other and transform the
-        data, then use fit_transform on transformed data using the final
+        """Fit the model and transform with the final estimator
```

Comment:
The tricky thing is that because these have legacy parameter validation in `__init__`, testing all validation cases is some effort!